File: my-folder/0040-combination-sum-ii/solution.ts
function combinationSum2(candidates: number[], target: number): number[][] {
    candidates.sort()
    let result = []
    backtrack(0, [], candidates, 0, target, result)
    return result
};

function backtrack(i: number, path: number[], candidates: number[], total: number, target: number, result: number[][]) {
    if (total === target) {
        result.push([...path])
        return
    }

    if (i >= candidates.length || total > target)
        return

    path.push(candidates[i])
    backtrack(i+1, path, candidates, total+candidates[i], target, result)

    while (i+1 < candidates.length && candidates[i]==candidates[i+1])
        i++

    path.pop()
    backtrack(i+1, path, candidates, total, target, result)
}

File: my-folder/0040-combination-sum-ii/solution.go
func combinationSum2(candidates []int, target int) [][]int {
    sort.Ints(candidates)
    var res [][]int
    backtrack([]int{}, target, 0, candidates, &res)
    return res
}

func backtrack(path []int, remain int, start int, candidates []int, res *[][]int) {
    if remain == 0 {
        tmp := make([]int, len(path))
        copy(tmp, path)
        *res = append(*res, tmp)
        return
    }

    if remain < 0 {
        return
    }

    for i := start; i<len(candidates); i++ {
        if i > start && candidates[i] == candidates[i-1] {
            continue
        }

        if candidates[i] > remain {
            break
        }

        backtrack(append(path, candidates[i]), remain-candidates[i], i+1, candidates, res)
    }
}
